### ISSUE TITLE
Ring Games

### ISSUE THREAD

**Issue by ctm** [ctm/mb2-doc#88]
Allow players to create ring games, which initially can  basically be events that show up and automatically start once two people commit some of their play money (implemented as part of #87).  Then allow additional players to join simply by picking an empty seat until the table is full.

More advanced features, like a sign-up with a minimum number of players and must-move tables can come later. In fact, the initial support can even skip things like missed blinds, etc, as long as any corners that we cut are documented so that people understand what's happening and not think it's a bug.

This is fairly important to the unsupervised play-testing milestone because it will allow people to have fun even when there's too few people for a tournament.

**Comment by ctm** [comment]
Some useful info from C.Meck:

> I've never seen the Moving Button Rule used in a tournament setting, but some card rooms use it for cash games.  It's the most complicated scenario, refer to the link posted above.  Essentially it is possible for 3 blinds to be posted on the same hand.  I wouldn't recommend using it, it's mainly used for cash games in places like California where a drop of a fixed amount of money per hand is used rather than a rake based on percentage of the pot, and to make sure players don't get to weasel out of paying their blinds if they step away for a few hands.  It's very confusing if you are not used to it--I've only played a few times in my life with this.

...

> Rather than the Moving Button Rule (which I've seen used in some rooms in California for cash games), what is more common for dealing with players missing the blinds in cash games is the following.

> Suppose the blinds are 1/2.  I'm absent from the table for both of my blinds.  The blinds aren't paid, and the dealer puts a Missed Blind button in front of me.  When I return to the table, I can either pay the blinds that I missed OR sit out until the next time I would be the big blind.  If I choose to post the blinds I missed, then you have the two players naturally in the blinds paying them, my $2 BB is posted in as if I'm limping into a hand, and I'll have the option to check or bet if no one raises before me, and my $1 SB is posted as 'dead money' (i.e. immediately pulled into the middle).  If I had only missed the small blind, then it is posted as 'dead money'.  

> "Buying the button" is a rule commonly used in cash games if you had just missed paying your big blind.  Instead of either having to wait to post your missed blinds until after the button passes you OR waiting an entire orbit until you would be in the BB again, you post both blinds on the hand that you would have been the SB and thus keep the right to have the button on the following hand (thus the name "buying the button").
https://www.tommyangelo.com/buy-button-perfect-rule/

**Comment by ctm** [comment]
I'm going to try to get ring games (even if they aren't polished) ready for vATLARGE. There's a lot of non-programming going on in my household right now that may interfere, but one needs goals.

**Comment by ctm** [comment]
I didn't get ring games going in time for vATLARGE. Once that deadline had past, I worked on a few other things before returning to this issue. I've been working on this off and on for a while and I have made some progress, although it won't look like it from the outside.

Internally, there's a fair amount of knowledge about the differences between tournaments and ring games, but there are still a ton of places where that knowledge isn't used. 

The good news is I can manually create a ring game in the database and have it automatically start when N players sign up. This means "all the plumbing" is in place, even though it's not always appropriately used. I do not currently have a complete list of places where we have to do things differently, but here are a few off the top of my head:

- [ ] When quitting, players need their `ChipCount` chips converted to `FunMoney`
- [ ] Getting down to fewer than two players should pause the table
- [ ] Need a way to join an existing ring game
- [ ] GameKeeper needs to know how to cycle through a `Vec<NGamesWithStakes>`
- [ ] When signing up, players need to be able to specify their buy-in
- [ ] Need a minimal UI for creating a ring game
- [ ] Need a database representation of a ring game structure
- [ ] Need. to enable rebuys
- [ ] Various stats that are collected need to be done with the ring game columns for ring games
- [ ] Need a waitlist

Eventually we'll also have to deal with esoterica like preventing ratholing, missed blinds, seat moves, etc. but the above are even higher priority. 

Of the above, the database representation of a ring game structure is sort of the hardest. In some sense it's drop dead simple, the exception being that we don't yet have a database representation of a game and the options that affect a game. That requires some architectural decisions that need to be deferred until we know how we're funded. OTOH, in the meantime, I can hard-code in few combinations of games and stakes so that I have something to work with as I finish the above todo items.

**Comment by ctm** [comment]
Yesterday I resumed working on this. As I mentioned in my [comment above](https://github.com/ctm/mb2-doc/issues/88#issuecomment-760542917), I chose to hard-code a combination of games and stakes and realized that I can deploy using that hack and I can also deploy without a waitlist. So, those two don't need to be done before I deploy and close this issue. They'll get their own issues at some point.

Locally I can create a ring game. I haven't actually tested it yet, because I got to that point at the end of my work day yesterday, but this morning I'll give it a spin. If there are no major surprises, I should be able to much through a bunch of the remaining check-boxes today.

Whee!

**Comment by ctm** [comment]
So, `GameKeeper` wasn't doing the right thing. Then I find this code:
```
    // Just like starting_chips, this goes away after more refactoring,
    // because a ring game doesn't have a level, but that requires a little
    // more refatoring than I'm doing tonight.
    pub fn first_level(&self) -> Level {
        use State::*;
        match self {
            Tournament(t) => t.structure.rounds[0].level.clone(),
            Ring(r) => r.structure.to_level(),
        }
    }
```
Oops. Yup, I'll have to refactor that out.

**Comment by ctm** [comment]
Although I haven't updated this issue recently, I've made steady progress every day for the last week. I'm starting ring games on my development system and playing, quitting and joining, but there are various rough edges that I'm smoothing out, and doing so without making check-boxes in this issue for me to fill in. So, from the outside it may appear that I'm not doing much, but I'm getting there.

**Comment by ctm** [comment]
I've updated my checkboxes to reflect what I need to do before I deploy. I've moved other, fairly important, features to their own issues and created the [`ring games`](https://github.com/ctm/mb2-doc/labels/ring%20games) label to help me keep track of them, since few (none?) of them will be tagged high priority.

**Comment by ctm** [comment]
Ugh. I just had a panic.

I started with two players and then added a third. The sit-out code caused the incoming player to sit out (which may have been correct, but it didn't look correct) and then on the next hand it caused one more player to sit out, which then meant we started the hand with one player and that panicked.

I am not surprised that the sit-out code is wrong, but I figured it would at most be overly aggressive and cause players to sit out when they shouldn't need to. Currently mb2 uses a dead button for both tournaments and ring games, only because I haven't implemented the forward moving button rule. So, to get everything perfectly correct will take a little work, although to get this to a point where I'm willing to deploy it I just need to make sure we cant get the panics.
